File: src/main/java/net/katsstuff/danmakucore/shape/ShapeHandler.java
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.shape;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import com.google.common.collect.ImmutableSet;

import net.katsstuff.danmakucore.data.Quat;
import net.katsstuff.danmakucore.data.Vector3;
import net.katsstuff.danmakucore.entity.danmaku.EntityDanmaku;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityLivingBase;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;
import net.minecraftforge.fml.common.gameevent.TickEvent;

public final class ShapeHandler {

	private static final List<IShapeEntry> shapeList = new ArrayList<>();

	/**
	 * Creates a new shape with a position as an anchor
	 * @return A set that will contain all the danmaku spawned by the shape. The set's content will change over time
	 */
	@SuppressWarnings("WeakerAccess")
	public static Set<EntityDanmaku> createShape(IShape shape, Vector3 pos, Quat orientation) {
		return createEntry(new ShapeEntryPosition(shape, pos, orientation));
	}

	/**
	 * Creates a new shape with an entity as an anchor
	 * @return A set that will contain all the danmaku spawned by the shape. The set's content will change over time
	 */
	@SuppressWarnings("UnusedReturnValue")
	public static Set<EntityDanmaku> createShape(IShape shape, Entity anchor) {
		return createEntry(new ShapeEntryEntity(shape, anchor));
	}

	/**
	 * Creates a new shape with a living entity(eye height) as an anchor.
	 * @return A set that will contain all the danmaku spawned by the shape. The set's content will change over time
	 */
	@SuppressWarnings("unused")
	public static Set<EntityDanmaku> createShape(IShape shape, EntityLivingBase anchor) {
		return createEntry(new ShapeEntryEntityLiving(shape, anchor));
	}

	/**
	 * Creates a new shape with a from the specific {@link ShapeEntry}
	 * @return A set that will contain all the danmaku spawned by the shape. The set's content will change over time
	 */
	@SuppressWarnings("WeakerAccess")
	public static Set<EntityDanmaku> createEntry(IShapeEntry entry) {
		shapeList.add(entry);
		return entry.getDrawn();
	}

	@SubscribeEvent
	public static void onTick(TickEvent.WorldTickEvent event) {
		if(event.phase == TickEvent.Phase.START) {
			List<IShapeEntry> completedShapes = shapeList.stream().filter(IShapeEntry::draw).collect(Collectors.toList());
			shapeList.removeAll(completedShapes);
		}
	}

	@SuppressWarnings("WeakerAccess")
	public interface IShapeEntry {

		/**
		 * Gets the current drawn danmaku from this shape. This set should always be the same reference.
		 */
		Set<EntityDanmaku> getDrawn();

		/**
		 * Draws this shape, giving it the information it needs.
		 * @return If this shape is completed and should be removed.
		 */
		boolean draw();
	}

	/**
	 * A class that specifies how a shape should be drawn.
	 */
	private abstract static class ShapeEntry implements IShapeEntry {

		final Set<EntityDanmaku> drawn = new HashSet<>();
		final IShape shape;
		int counter = 0;

		ShapeEntry(IShape shape) {
			this.shape = shape;
		}

		@Override
		public Set<EntityDanmaku> getDrawn() {
			return drawn;
		}
	}

	private abstract static class ShapeEntryDynamicPos<T> extends ShapeEntry {

		@SuppressWarnings("WeakerAccess")
		protected final T dynPos;

		ShapeEntryDynamicPos(IShape shape, T dynPos) {
			super(shape);
			this.dynPos = dynPos;
		}

		@Override
		public boolean draw() {
			Vector3 pos = getCurrentPos();
			Quat orientation = getCurrentOrientation();

			ShapeResult ret = shape.drawForTick(pos, orientation, counter);
			drawn.addAll(ret.getSpawnedDanmaku());
			shape.doEffects(pos, orientation, counter, ret.getSpawnedDanmaku(), ImmutableSet.copyOf(drawn));

			counter++;
			return ret.isDone();
		}

		protected abstract Vector3 getCurrentPos();
		protected abstract Quat getCurrentOrientation();
	}

	private static class ShapeEntryEntity extends ShapeEntryDynamicPos<Entity> {

		ShapeEntryEntity(IShape shape, Entity dynPos) {
			super(shape, dynPos);
		}

		@Override
		protected Vector3 getCurrentPos() {
			return new Vector3(dynPos);
		}

		@Override
		protected Quat getCurrentOrientation() {
			return Quat.orientationOf(dynPos);
		}
	}

	private static class ShapeEntryEntityLiving extends ShapeEntryDynamicPos<EntityLivingBase> {

		ShapeEntryEntityLiving(IShape shape, EntityLivingBase dynPos) {
			super(shape, dynPos);
		}

		@Override
		protected Vector3 getCurrentPos() {
			return new Vector3(dynPos);
		}

		@Override
		protected Quat getCurrentOrientation() {
			return Quat.orientationOf(dynPos);
		}
	}

	private static class ShapeEntryPosition extends ShapeEntry {

		private final Vector3 pos;
		private final Quat orientation;

		ShapeEntryPosition(IShape shape, Vector3 pos, Quat orientation) {
			super(shape);
			this.pos = pos;
			this.orientation = orientation;
		}

		@Override
		public boolean draw() {
			ShapeResult ret = shape.drawForTick(pos, orientation, counter);
			drawn.addAll(ret.getSpawnedDanmaku());
			shape.doEffects(pos, orientation, counter, ret.getSpawnedDanmaku(), ImmutableSet.copyOf(drawn));

			counter++;
			return ret.isDone();
		}
	}
}

File: src/main/scala/net/katsstuff/danmakucore/shape/ShapeHandler.scala
/*
 * This class was created by <Katrix>. It's distributed as
 * part of the DanmakuCore Mod. Get the Source Code in github:
 * https://github.com/Katrix-/DanmakuCore
 *
 * DanmakuCore is Open Source and distributed under the
 * the DanmakuCore license: https://github.com/Katrix-/DanmakuCore/blob/master/LICENSE.md
 */
package net.katsstuff.danmakucore.shape

import scala.concurrent.{Future, Promise}

import net.katsstuff.teamnightclipse.mirror.data.{Quat, Vector3}
import net.minecraft.entity.{Entity, EntityLivingBase}
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.gameevent.TickEvent

object ShapeHandler {
  private var shapeList = Seq.empty[(IShapeEntry, Promise[ShapeResult])]

  /**
    * Creates a new shape with a position as an anchor
    *
    * @return A set that will contain all the danmaku spawned by the shape. The set's content will change over time
    */
  def createShape(shape: Shape, pos: Vector3, orientation: Quat): Future[ShapeResult] =
    createEntry(ShapeEntryPosition(shape, pos, orientation, 0, Set.empty))

  /**
    * Creates a new shape with an entity as an anchor
    *
    * @return A set that will contain all the danmaku spawned by the shape. The set's content will change over time
    */
  def createShape(shape: Shape, anchor: Entity): Future[ShapeResult] =
    createEntry(ShapeEntryEntity(shape, anchor, 0, Set.empty))

  /**
    * Creates a new shape with a living entity(eye height) as an anchor.
    *
    * @return A set that will contain all the danmaku spawned by the shape. The set's content will change over time
    */
  def createShape(shape: Shape, anchor: EntityLivingBase): Future[ShapeResult] =
    createEntry(ShapeEntryEntityLiving(shape, anchor, 0, Set.empty))

  /**
    * Creates a new shape with a from the specific [[ShapeEntry]]
    *
    * @return A set that will contain all the danmaku spawned by the shape. The set's content will change over time
    */
  def createEntry(entry: IShapeEntry): Future[ShapeResult] = {
    val promise = Promise[ShapeResult]
    shapeList = shapeList :+ (entry -> promise)
    promise.future
  }

  @SubscribeEvent
  def onTick(event: TickEvent.WorldTickEvent): Unit = if (event.phase == TickEvent.Phase.START) {
    shapeList = shapeList.flatMap {
      case (entry, promise) =>
        val (result, next) = entry.draw
        promise.success(result)

        for {
          n           <- next
          nextPromise <- result.promise
        } yield n -> nextPromise
    }
  }

  trait IShapeEntry {

    /**
      * Draws this shape, giving it the information it needs.
      *
      * @return If this shape is completed and should be removed.
      */
    def draw: (ShapeResult, Option[IShapeEntry])
  }

  /**
    * A class that specifies how a shape should be drawn.
    */
  trait ShapeEntry extends IShapeEntry {
    def shape: Shape

    def allDrawn: Set[ShapeResult]
    def tick: Int
  }

  abstract private class ShapeEntryDynamicPos[T] extends ShapeEntry {
    def dynPos: T

    override def draw: (ShapeResult, Option[IShapeEntry]) = {
      val pos         = getCurrentPos
      val orientation = getCurrentOrientation
      val ret         = shape.draw(pos, orientation, tick)

      shape.doEffects(pos, orientation, tick, ret, allDrawn)
      if (ret.isDone) (ret, None) else (ret, Some(create(tick + 1, allDrawn + ret)))
    }

    def create(tick: Int, allDrawn: Set[ShapeResult]): IShapeEntry

    protected def getCurrentPos: Vector3
    protected def getCurrentOrientation: Quat
  }

  private case class ShapeEntryEntity(shape: Shape, dynPos: Entity, tick: Int, allDrawn: Set[ShapeResult])
      extends ShapeEntryDynamicPos[Entity] {
    override protected def getCurrentPos                                    = new Vector3(dynPos)
    override protected def getCurrentOrientation: Quat                      = Quat.orientationOf(dynPos)
    override def create(tick: Int, allDrawn: Set[ShapeResult]): IShapeEntry = copy(tick = tick, allDrawn = allDrawn)
  }

  private case class ShapeEntryEntityLiving(
      shape: Shape,
      dynPos: EntityLivingBase,
      tick: Int,
      allDrawn: Set[ShapeResult]
  ) extends ShapeEntryDynamicPos[EntityLivingBase] {
    override protected def getCurrentPos                                    = new Vector3(dynPos)
    override protected def getCurrentOrientation: Quat                      = Quat.orientationOf(dynPos)
    override def create(tick: Int, allDrawn: Set[ShapeResult]): IShapeEntry = copy(tick = tick, allDrawn = allDrawn)
  }

  private case class ShapeEntryPosition(
      shape: Shape,
      pos: Vector3,
      orientation: Quat,
      tick: Int,
      allDrawn: Set[ShapeResult]
  ) extends ShapeEntry {
    override def draw: (ShapeResult, Option[IShapeEntry]) = {
      val ret = shape.draw(pos, orientation, tick)
      shape.doEffects(pos, orientation, tick, ret, allDrawn)
      if (ret.isDone) (ret, None) else (ret, Some(copy(tick = tick + 1, allDrawn = allDrawn + ret)))
    }
  }
}
